### PR TITLE
[Feature] Allow RelationManger to create secondary for HasOne/BelongsTo relationships

### PR DESCRIPTION
**Pull request by mallardduck**
Per the discussion I started here: https://github.com/laravel-filament/filament/discussions/319

This PR is a quick and simple version of the concept that I worked up with today.
There is one minor refactor to the CallsHooks interface so it can be reused better.
The rest of the changes add a simple implementation based on the concepts used in the RelationshipManagers

This is a working implementation and will play well with at least `hasOne` relations.
Here is the code I've used to test the impelementation to give a sense of user side API - basically same as ResourceManagers:

```php
<?php

namespace App\Blog\Filament\Resources\BlogPostResource\RelationManagers;

use Filament\Resources\Forms\Form;
use Filament\Resources\Forms\Components;
use Filament\Resources\SubFormRelationManager; // NOTE: this is just an abstract base class that extends RelationManager

class SeoDataRelationshipManager extends SubFormRelationManager
{
    public static $relationship = 'seoData';
    public static $title = 'SEO Data';

    public static function form(Form $form)
    {
        return $form->schema([
            Components\TextInput::make('title')
                ->label("Open Graph Title")
                ->placeholder('A title for use in the open-graph metadata.'),
            Components\Textarea::make('description')
                ->label("Open Graph Description")
                ->placeholder('A description for use in the open-graph metadata.'),
        ]);
    }
}
```

Then on the `BlogPostResource` you'd do:
```php
    public static function relations()
    {
        return [
            RelationManagers\SeoDataRelationshipManager::class,
            RelationManagers\SecondaryCategoriesRelationshipManager::class,
        ];
    }
```

### CODE REVIEW

**Comment by danharrin** [comment]
So as far as I can see, this is basically a `BelongsTo` / `HasOne` relation manager, right? Very cool. I think we should implement it inside the existing relation manager though. If the relation is either `BelongsTo` or `HasMany`, this form should render instead of the table. What do you think?

**Comment by mallardduck** [comment]
Yeah - I think this is a pretty fair way to look at it:
> this is basically a `BelongsTo` / `HasOne` relation manager

> we should implement it inside the existing relation manager
From a cursory view of things I couldn't find a method to support that. However I do like the idea of keeping it within the same framework as the relation manager. It is effectively the same concern and results, but with slightly different context.

>  If the relation is either `BelongsTo` or `HasOne`, this form should render instead of the table.

yeah - I really like that idea. in the RFC thread I mentioned the potential of inlining the fields into the main form. I do think that would make a great default UX. so I think this all sounds great!

I'll see what I can do about implementing this as a feature of the relation manager and update the PR once I'm done for more feedback.

**Comment by mallardduck** [comment]
Hey @danharrin - so I could use some information on how I can get this working fully. By adjusting the PR technically things are being done via `RelationshipManager` now. However the form is not loading the data from the correct model.

Currently the visual results are identical to those in my [screenshot here](https://user-images.githubusercontent.com/619938/113528837-d9b03d80-958f-11eb-92ed-e58d3f160b43.png) - except with no form fields filled in. I suspect this will require using the `switchRecord` method, however I'm woefully ignorant on how that should be done.

I'm likely overthinking it - but I also assumed changes may need to be made to `switchRecord` as it seems to rely on using a call to Eloquent's `find` - where as with this type of interaction it may be simpler/better to just pass the record directly for `HasOne` and `BelongsTo`.

I'm also a bit unsure where that switch should happen at - I did however play around and get something working. So that version of the PR is pushed now. However it just seems off and I'm not sure how to correct it.

File: src/Resources/RelationManager/EditRecord.php [hunk 1]
```diff
@@ -52,6 +52,8 @@ public function mount()
 
     public function render()
     {
+        // TODO: figure out how to conditionally fire this for HasOne and BelongsTo
+        $this->switchRecordInstance($this->getRelationship()->first());
```
Comment (mallardduck) [inline src/Resources/RelationManager/EditRecord.php:56]: @danharrin - note the TODO, as the manager property in this class isn't an instance we cannot use:

```php
if (
    $this->isType(Relations\HasOne::class) ||
    $this->isType(Relations\BelongsTo::class)
)
```

As I would have planned.

File: src/Resources/RelationManager/EditRecord.php [hunk 2]
```diff
@@ -99,6 +101,16 @@ public function switchRecord($manager, $recordKey)
         $this->callHook('afterFill');
     }
 
+    public function switchRecordInstance($record)
```
Comment (mallardduck) [inline src/Resources/RelationManager/EditRecord.php:104]: And as mentioned I'm not sure what the best or most logical way to acquire the proper record/model is.
I'm sure that i am overlooking a more simple way to do this that's already supported.

**Comment by mallardduck** [comment]
Additionally - the current PR is still rendering the fields as a secondary form out of ease.
Given that I ran into the issue getting the proper field values, I didn't want to work on integrating the fields into the main form as I assumed that would be much harder - and solvable after I understand the proper way to swap records.

File: src/Resources/SubFormRelationManager.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Filament\Resources;
+
+/**
+ * This is a stub class for HasOne/BelongsTo relationships to add forms on a resource page.
+ * It will work largely like the RelationManager, just without the table method.
+ */
+abstract class SubFormRelationManager extends RelationManager
```
Comment (mallardduck) [inline src/Resources/SubFormRelationManager.php:9]: The benefit of keeping it a novel abstract class is that it prevents a need to add more conditional checks to the `canCrate`, `canAttach` and `canDetach` methods. It also prevents the need for an always empty table method that's required due to extending the `RelationManager`.

Both of these are a benefit as this use case will never need Attach/Detach - nor do we want to invoke any thing related to the default `RelationManager` create process. We just want the EditRecord option essentially so we can embed it into the page.

**Comment by mallardduck** [comment]
@telkins - this PR was only ever prepared to work with HasOne/BelongsTo relationships.
In theory it could potentially support BelongsToMany, however I don't believe there are any plans to actually use this PR at this point. So I wouldn't hold my breath on it TBH, tho maybe if more community support and requests for the feature come maybe it'll be a priority.

**Comment by telkins** [comment]
> @telkins - this PR was only ever prepared to work with HasOne/BelongsTo relationships.

@mallardduck Thanks for the info/answer. 👍 

> In theory it could potentially support BelongsToMany, however I don't believe there are any plans to actually use this PR at this point.

Why do you believe this PR won't be used? 🤔 

> So I wouldn't hold my breath on it TBH, tho maybe if more community support and requests for the feature come maybe it'll be a priority.

That's unfortunate.  I've only used it a bit and it seems quite nice.  It's very clean and fast and seems to be well thought out.  However, I hit a bump pretty quickly, and that's not normally a good sign.  I hope it continues to get some dev love.  I'd really like to see this become successful. 🤓

**Comment by mallardduck** [comment]
> Why do you believe this PR won't be used? 🤔

I guess it's probably wrong to imply it won't be used at all - but rather it's been tagged to be held off on until V2 which is TBD and doesn't (AFAIK) have a firm roadmap and timeline for when to expect it. TBH that lack of clarity is why I've stopped providing PRs as rapidly. I like this project as a solution for a lot of issues, but I also like to move fast and can't be putting projects on hold for an eventual V2 release to provide the features I need and want today.

Especially when I'm putting in the leg work to make the PRs - it makes it sting more when a projects progress needs to stagnate since I already put the effort in but don't get benefit of the results. Unfortunately because of that doing things in a way where I cannot more easily share them, but I get the benefits has been where I've put my efforts lately.

**Comment by telkins** [comment]
> > Why do you believe this PR won't be used? 🤔
> 
> I guess it's probably wrong to imply it won't be used at all - but rather it's been tagged to be held off on until V2 which is TBD and doesn't (AFAIK) have a firm roadmap and timeline for when to expect it.

Wow...!  I didn't even realize they were working on a V2.  That's good to know.

> TBH that lack of clarity is why I've stopped providing PRs as rapidly. I like this project as a solution for a lot of issues, but I also like to move fast and can't be putting projects on hold for an eventual V2 release to provide the features I need and want today.

I hear you.  It's frustrating to wait on fixes/features, but considerably more so when you (or someone else) has put in the work for a viable PR.  Sorry to hear about that.

> Especially when I'm putting in the leg work to make the PRs - it makes it sting more when a projects progress needs to stagnate since I already put the effort in but don't get benefit of the results. Unfortunately because of that doing things in a way where I cannot more easily share them - but I get the benefits - has been where I've put my efforts lately.

Understandable.  Perhaps you could have a (private) chat with one or more of the maintainers to see if you could find a way to help with the forward momentum...?  I suspect they're busy with other stuff.  I know I've got a lot of irons in the fire.  I have nowhere near the number of people interested in the repos I maintain, but sometimes it's hard to devote the time/attention that I would like to see.  Not sure if that makes sense, but I hope you know what I'm saying.

Anyway, give it a try, if you haven't already.  It could be good for everyone.  🤓

**Comment by danharrin** [comment]
Hey @mallardduck @telkins just so you're know, I'm aware I've been more inactive recently - I've had a period of exams and I'm graduating next week. Momentum should be picked up again from that point onwards :)

v2 will be built in public, so you'll be able to monitor the feature progress as it happens. We'll probably aim for a Summer Laracon release or thereabouts 👍

**Comment by mallardduck** [comment]
Hey @danharrin - no worries at all and hope my comments didn't come off as too harsh. Most of us have 'day lives' outside of open source so I understand when periods of low churn on projects happen.

To be clear in my case I got into using filament just for fun making a blog to see how far I could go with it. Then I decided to start using it with a project for work which required a higher velocity of progress. So since I noticed that lower pace I simply opted to just start hacking things I need into place in my project directly. The placeholder PR I sent over is an example of that - something I initially developed within filament but have ported to my app locally.

In theory hopefully some of these other things I'm doing can benefit the wider community when V2 work starts up. Assuming these solutions are relevant still since as you mentioned in a different PR - there have been cases where your team has plans to implement features providing similar results via different APIs.

So to be clear it's not as though I've given up on helping provide contributions to filament overall - I'm just having to be more pragmatic about where I choose to implement something. To @telkins point though, I wouldn't mind pushing things forward for filament when you and the others don't have the time and bandwidth. However I know that there's a "cost" of delegating tasks which usually comes in the form of lots of docs. And it's not like that doesn't take time too - so sometimes it's not quite as simple to accept that kind of offer for help either.

Either way, however things go I'll be here for it - and will help as I can.

**Comment by danharrin** [comment]
Yeah thank you, I totally understand how you're feeling.

v2 will likely be a complete rewrite, which makes me hesitant to merge existing PRs until I've got a plan on how features will be implemented 👍

Anyway, as I've mentioned before, thanks for all your contributions to the project! Even if they don't get merged, we can still grab code from them which helps us loads.

**Comment by mallardduck** [comment]
@danharrin - As I just started working with V2 recently I've found that this area is still lacking. For instance, in a table I can do this:
```php
                Tables\Columns\BooleanColumn::make('crawlStatus.is_active')
                    ->label('Status')->sortable(),
```

However in a form I cannot do:

```php
                Forms\Components\Checkbox::make('crawlStatus.is_active')
                    ->label('Crawl Active'),
```

Am I missing something in the docs that explains how to use relationship fields like this? Or has this feature not been implemented in V2 yet - if so should I work on a new version of this PR's concept?

**Comment by danharrin** [comment]
Hey! You can achieve this using:

```php
Forms\Components\Checkbox::make('crawlStatus.is_active')
    ->label('Crawl Active')
    ->saveRelationshipsUsing(fn ($record, $state) => $record->crawlStatus->update(['is_active' => $state]))
    ->dehydrated(false),
```

Very soon, I want people to be able to wrap their form components in a BelongsTo or MorphOne group which handles this all for you, should not be too tricky :)

**Comment by mallardduck** [comment]
Oh wow! Very cool - I must have completely missed that in docs and hadn't dug enough via IDE to discover that.
Thank you!

**Comment by danharrin** [comment]
It's just an internal API for the moment, but I'll publicise it when I make those relationship groups :)

**Pull request closed without merging**